alternator/server.cc

```diff
@@ -211,9 +212,28 @@ class local_nodelist_handler : public gated_handler {
         // using _gossiper().get_live_members(). But getting
         // just the list of live nodes in this DC needs more elaborate code:
         auto& topology = _proxy.get_token_metadata_ptr()->get_topology();
-        sstring local_dc = topology.get_datacenter();
-        std::unordered_set<gms::inet_address> local_dc_nodes = topology.get_datacenter_endpoints().at(local_dc);
+        // /localnodes lists nodes in a single DC. By default the DC of this
+        // server is used, but it can be overriden by a "dc" query option.
+        // If the DC does not exist, we return an empty list - not an error.
+        sstring query_dc = req->get_query_param("dc");
+        sstring local_dc = query_dc.empty() ? topology.get_datacenter() : query_dc;
+        std::unordered_set<gms::inet_address> local_dc_nodes;
+        const auto& endpoints = topology.get_datacenter_endpoints();
+        auto dc_it = endpoints.find(local_dc);
+        if (dc_it != endpoints.end()) {
+            local_dc_nodes = dc_it->second;
+        }
+        // By default, /localnodes lists the nodes of all racks in the given
+        // DC, unless a single rack is selected by the "rack" query option.
+        // If the rack does not exist, we return an empty list - not an error.
+        sstring query_rack = req->get_query_param("rack");
```

Comment:
Don't we want to check that values of `dc` & `rack` parameters are even valid? For example if somebody inserted into the URL: `?dc=-1`, which doesn't make any sense, we'd just return nothing.

Reply:
I don't know if "dc=-1" doesn't make sense - it just refers to a DC named "-1" - is this an illegal name?
See my discussion above about what to do in case of a DC or rack which doesn't "exist", i.e., doesn't have any nodes in it. There are pros and cons to both returning an empty list and returning an HTTP error, and I am honestly not sure which is "better", so I went with the one that made a tiny bit more sense to me.

Reply:
OK, me saying "`?dc=-1` doesn't make any sense" is a bit of an overstatement. If we don't have any way of knowing if the DC name is legal, meaning - it matches some dc-naming-pattern, I guess  we can resolve this discussion.

Reply:
Another topic, do we want to make sure that when `rack` is provided it goes along with `dc`?
Currently `?rack=my_rack` is valid request, which potentially can give user different results, depedning on which dc request lands.

Reply:
@dkropachev I actually thought it makes sense. If you already know a seed node inside the right DC, you can use it to get other nodes in your rack, without needed to know the name of the DC your seed node is in.

Reply:
Ok